recipes/openai_textcat.py

```diff
@@ -0,0 +1,283 @@
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Callable, Dict, Iterable, List, Optional
+
+import prodigy
+import spacy
+import srsly
+import tqdm
+from dotenv import load_dotenv
+from jinja2 import Template
+from prodigy.util import msg
+from spacy.language import Language
+
+from recipes.openai import OpenAISuggester, PromptExample, get_api_credentials
+from recipes.openai import load_template, normalize_label
+from recipes.openai import read_prompt_examples
+
+CSS_FILE_PATH = Path(__file__).parent / "style.css"
+DEFAULT_PROMPT_PATH = (
+    Path(__file__).parent.parent / "templates" / "textcat_prompt.jinja2"
+)
+HTML_TEMPLATE = """
+<div class="cleaned">
+  {{ #label }}
+    <centering>
+    <h2>OpenAI ChatGPT says: {{ meta.answer }}</h2>
+    </centering>
+  {{ /label }}
+  <details>
+    <summary>Show the prompt for OpenAI</summary>
+    <pre>{{openai.prompt}}</pre>
+  </details>
+  <details>
+    <summary>Show the response from OpenAI</summary>
+    <pre>{{openai.response}}</pre>
+  </details>
+</div>
+"""
+
+# Set up openai
+load_dotenv()  # take environment variables from .env.
+
+
+@dataclass
+class TextCatPromptExample(PromptExample):
+    """An example to be passed into an OpenAI TextCat prompt."""
+
+    text: str
+    answer: str
+    reason: str
+
+    @classmethod
+    def from_prodigy(cls, example: Dict, labels: Iterable[str]) -> "PromptExample":
+        """Create a prompt example from Prodigy's format."""
+        if "text" not in example:
+            raise ValueError("Cannot make PromptExample without text")
+
+        full_text = example["text"]
+        reason = example["meta"].get("reason")
+        if len(labels) == 1:
+            answer = "accept" if example.get("answer") else "reject"
+        else:
+            answer = example.get("accept")
+        return cls(text=full_text, answer=answer, reason=reason)
+
+
+def make_textcat_response_parser(labels: List[str]) -> Callable:
```

Comment:
The parser implementation is now a function instead of inheriting some base class. So we have a `make_textcat_response_parser` return a function that performs the parsing. 

I'd want to do the same pattern with `PromptExample`, but still figuring out the best way to do so.

Reply:
I'm a bit stuck on doing the same thing with `PromptExample`. There is the [`make_dataclass` function](https://docs.python.org/3/library/dataclasses.html#dataclasses.make_dataclass) that make the whole thing more dynamic but I'm still wrapping around how to elegantly phrase the function. I'm wary of overcomplicating it. Will get back to this.

Reply:
Do we want these recipes to make use of the `StructuredExample` in v1.12? That sounds reasonable, but it also implies that we can wait until that feature has progressed a bit.

The main thing we should document is that the contents of an example in a stream will be passed to the template in order to render them. From there, it's kind of the end users responsibility, no?

Reply:
> Do we want these recipes to make use of the StructuredExample in v1.12? That sounds reasonable, but it also implies that we can wait until that feature has progressed a bit.

That might be a good idea. I can keep the current iteration of `PromptExample` first and then document how to use it so we don't need to wait for `StructuredExamples` for this to get merged, what do you think?

Reply:
Documented PromptExample usage here: https://github.com/explosion/prodigy-openai-recipes/pull/54/commits/405a243aa4b89c7e60aaad59f0f4e01b972b5473